src/main/scala/dataduggee/DataDuggee.scala

```diff
@@ -69,32 +68,40 @@ object DataDuggee {
         maxMetrics: Int = 1024,
         maxDelay: FiniteDuration = 10.seconds,
         maxConcurrrency: Int = 512
-    ): Pipe[F, Metric, Unit] = { in =>
+    ): Pipe[F, Metric, String] = { in =>
       in.groupWithin(maxMetrics, maxDelay)
         .mapAsync(maxConcurrrency) { xs =>
-          xs.toNel.fold(().pure[F])(sendMetrics)
+          xs.toNel.fold("".pure[F])(sendMetrics)
         }
     }
 
-    def sendMetrics(metrics: NonEmptyList[Metric]): F[Unit] = {
-
+    def sendMetrics(metrics: NonEmptyList[Metric]): F[String] = {
       val body: Stream[F, String] = codec.encodeMetrics(Stream.emits(metrics.toList))
-
-      val response = for {
+      for {
         req <- POST(body.through(fs2.text.utf8Encode), postMetricsUri, `Content-Type`(MediaType.application.json))
-        response <- client.successful(req)
+        response <- client.expectOr[String](req)(parseError)
       } yield response
-
-      response.void
     }
 
-    def createEvent(event: Event): F[Unit] = {
+    def createEvent(event: Event): F[String] = {
       val body = codec.encodeEvent(event)
-      val response = for {
+      for {
         req <- POST(body, postEventUri, `Content-Type`(MediaType.application.json))
-        resp <- client.successful(req)
-      } yield resp
-      response.void
+        response <- client.expectOr[String](req)(parseError)
+      } yield response
+    }
+
+    // TODO: Parse the DataDog error instead of just sending the whole JSON back as a String
+    private def parseError(resp: Response[F]): F[Throwable] = {
+      val code = resp.status.code
+      resp.body
+        .fold(new StringBuilder) { case (sb, b) => sb.append(b.toChar) }
+        .covary[F]
+        .compile
+        .lastOrError
+        .map { body =>
+          RequestError(code, body.toString)
+        }
```

Comment:
Let me check it out an alternative

Reply:
Yeah, I guess this is the only bit really left.

Reply:
`.body` is always a `Stream` I think, so unless you want to return a `Stream`, I don't think there is a way to avoid running the body